pandas/core/ops/array_ops.py

```diff
@@ -132,3 +152,177 @@ def na_arithmetic_op(left, right, op, str_rep, eval_kwargs):
         result = masked_arith_op(left, right, op)
 
     return missing.dispatch_fill_zeros(op, left, right, result)
+
+
+def arithmetic_op(left, right, op, str_rep, eval_kwargs):
+
+    from pandas.core.ops import (
+        maybe_upcast_for_op,
+        should_extension_dispatch,
+        dispatch_to_extension_op,
+    )
+
+    keep_null_freq = isinstance(
+        right,
+        (
+            ABCDatetimeIndex,
+            ABCDatetimeArray,
+            ABCTimedeltaIndex,
+            ABCTimedeltaArray,
+            Timestamp,
+        ),
+    )
+
+    # NB: We assume that extract_array has already been called on `left`, but
+    #  cannot make the same assumption about `right`.  This is because we need
+    #  to define `keep_null_freq` before calling extract_array on it.
+    lvalues = left
+    rvalues = extract_array(right, extract_numpy=True)
+
+    rvalues = maybe_upcast_for_op(rvalues, lvalues.shape)
+
+    if should_extension_dispatch(left, rvalues) or isinstance(
+        rvalues, (ABCTimedeltaArray, ABCDatetimeArray, Timestamp)
```

Comment:
Why is the check for rvalues being ABCTimedeltaArray or ABCDatetimeArray not incluced in should_extension_dispatch?

Reply:
historically should_extension_dispatch was more focused on dispatching to EA implementations, while the TDA/DTA/Timestamp check is specific to handling the `freq` attribute.  Combining them wouldn't be unreasonable.